File: queue/queue_circular.py
#https://www.programiz.com/dsa/circular-queue

class CircularQueue():

    def __init__(self, k):
        self.k = k
        self.queue = [None]*k
        self.head = self.tail = -1

    #insert and element in the queue
    def enqueue(self, data):

        if((self.tail + 1)%self.k == self.head):
            print("it is full")

        elif(self.head == -1):
            self.head = 0
            self.tail = 0
            self.queue[self.tail] = data
        else:
            self.tail = (self.tail + 1)%self.k
            self.queue[self.tail] = data
    
    #Delete an element 
    def dequeue(self):
        if(self.head == -1):
            print("Circular queue is empty")
        elif(self.head == self.tail):
            self.head = -1
            self.tail = -1

        else:
            temp = self.queue[self.head]
            self.head = (self.head + 1)%self.k

    def printQueue(self):
        if(self.head == -1):
            print("no element")
        
        if(self.tail >= self.head):
            for i in range(self.head, self.tail+1):
                print(str(self.queue[i]) + ", ")
        
        else:
            for i in range(self.head, self.k):
                print(str(self.queue[i]) + ", ")
            for i in range(0, self.tail + 1):
                print(str(self.queue[i] )+ ", ")


if __name__ == '__main__':
    obj = CircularQueue(5)
    obj.enqueue(1)
    obj.enqueue(2)
    obj.enqueue(3)
    obj.enqueue(4)
    obj.enqueue(5)
    obj.enqueue(6)
    print("Initial queue")
    obj.printQueue()
    obj.dequeue()
    print("After removing an element from the queue")
    obj.printQueue()

File: queue/queue_circular.cpp
//https://www.programiz.com/dsa/circular-queue


#include <iostream>
#define SIZE 5

using namespace std;

class Queue{

    private:
        int items[SIZE], front, rear;
    
    public:
    Queue(){
        front = -1;
        rear = -1;
    }

    bool isFull(){
        if( front == 0 && rear == SIZE -1 ){
            return true;
        }
        if( front == rear + 1){
            return true;
        }
        else
            return false;

    }

    bool isempty(){

        if(front == -1 && rear == -1){
            return true;
        }
        else
            return false;

    }

    //Adding an element
    void enqueue(int element){
        
        if(isFull()){
            cout<<endl;
            cout<<"it is full" << endl;
        }

        
        else{
            

            if(front == -1){
                front = 0;}
            
            rear = (rear + 1)%SIZE;
            cout << rear;
            items[rear] = element;
            
           


        }
    }

    //Removing an element
    void dequeue(){

        if(isempty()){
            cout<<" it is empty" << endl;
        }
        else{
            if(front == rear){
                front = -1;
                rear = -1;
            }
            else{
                front = (front + 1)%SIZE;
            }
        }

    }

    void display(){
        if(isempty()){
            cout<<"its empty" << endl;
        }

        else{
            for (int i = front; i != rear; i = (i + 1) % SIZE)
                cout << items[i];
        }
        cout<<endl;     
    }
};


int main() {
  Queue q;

 
  q.dequeue();
  q.enqueue(1);
  q.enqueue(2);
  q.enqueue(3);
  q.enqueue(4);
  q.enqueue(5);
  q.enqueue(6);
  q.display();
  q.dequeue();
  q.display();
  q.dequeue();
  
  return 0;
}